### ISSUE TITLE
Bug in Factor's reduce method

### ISSUE THREAD

**Issue by ankurankan** [pgmpy/pgmpy#222]
``` python3
f = Factor(['x1'], [2], [0.4, 0.6])
f.reduce('x1_0')
f.values
Out[32]: array([ 0.4])

f.variables
Out[33]: OrderedDict()

f.cardinality
Out[35]: array([], dtype=int64)
```

Expected Behavior:

``` python3
f = Factor(['x1'], [2], [0.4, 0.6])
f.reduce('x1_0')

f.values
Out: array([0.4])

f.variables
Out: OrderedDict([('x1', ['x1_0'])])

f.cardinality
Out: array([1])
```

**Comment by palashahuja** [comment]
@ankurankan expected behaviour for 

``` python3
f = Factor (['x1','x2','x3'], [2,3,2],np.arange (12))
f.reduce (['x1_0','x2_0','x3_0'])
```

**Comment by ankurankan** [comment]
@palashahuja 

``` python3
f.variables
Out: OrderedDict([('x1', ['x1_0']), ('x2', ['x2_0']), ('x3', ['x3_0'])])

f.cardinality
Out: array([1, 1, 1])

f.values
Out: array([0])
```

**Comment by palashahuja** [comment]
@ankurankan 
Could you please explain why do all the terms should not reduce to null value when they are reduced simultaneously. Instead,what we can do  is make some changes in the `_repr_` function  so that when the table doesn't contain any value, there should be a default error statement. Also, should I add the above two cases in the test cases ?

**Comment by abinashpanda** [comment]
@palashahuja Reduce means reducing the whole distribution to some context. For example when we reduce a factor f (as you have defined above) to the context of `x1_0`, we should be only having those entries where `x1` had assignment `0`. So if you reduce the whole table to the context of `[x1_0, x2_0, x3_0]`, the table should contain value corresponding to the assignment `x1 = 0`, `x2 = 0` and `x3 = 0`. It should not reduce to null. 

I hope this explanation helps. Let me know if you still have any doubt.

**Comment by ankurankan** [comment]
@palashahuja From the docstring of `reduce`: `Reduces the factor to the context of given variable values.`. So in the above example when we are trying to reduce the factor to `[x1_0, x2_0, x3_0]` we want to remove all the values from the factor that are not consistent with these states which leaves us with a single value in the factor. And this also reduces the cardinalities to `[1, 1, 1]` because for each variable we only have one state.

**Comment by palashahuja** [comment]
@ankurankan  Should I add the test cases also ?

**Comment by ankurankan** [comment]
@palashahuja Yes please add test cases for these if not present.

**Comment by palashahuja** [comment]
@ankurankan I think you should close this.

**Comment by ankurankan** [comment]
Fixed in #316. Behavior not changed left as previous because new behavior not consistent for single variable factors and multiple variable factors.
